src/main/java/baron/tasks/TaskManager.java

```diff
@@ -0,0 +1,271 @@
+     * @throws BaronException If there is no task, the specified task was not marked,
+     *                        or the index is out of bound.
+     */
+    public boolean unmarkTask(int index) throws BaronException {
+        if (this.isEmpty()) {
+            throw new BaronException(Message.MESSAGE_NO_TASK);
+        } else if (index > 0 && index <= this.getTaskCount()) {
+            if (this.taskList.get(index - 1).unmark()) {
+                return true;
+            } else {
+                throw new BaronException("The task is not marked.");
+            }
+        } else {
+            throw new BaronException(Message.generateTaskIndexOutOfBoundMessage(this.getTaskCount()));
+        }
+    }
+
+    /**
+     * Removes the task that is at the specified position (1-indexed).
+     * Shifts any subsequent tasks to the left.
+     *
+     * @param index the index of the task (1-indexed) to be removed.
+     * @return removed {@code Task} if it was removed successfully.
+     * @throws BaronException If there is no task or the index is out of bound.
+     */
+    public Task deleteTask(int index) throws BaronException {
+        if (this.isEmpty()) {
+            throw new BaronException(Message.MESSAGE_NO_TASK);
+        } else if (index > 0 && index <= this.getTaskCount()) {
+            this.previousTaskList = this.getAllTasks();
+            Task deletedTask = this.taskList.get(index - 1);
+            this.taskList.remove(index - 1);
+            return deletedTask;
+        } else {
+            throw new BaronException(Message.generateTaskIndexOutOfBoundMessage(this.getTaskCount()));
+        }
+    }
+
+    /**
+     * Filters the task list in this {@code TaskManager} by the specified {@code keyword}.
+     *
+     * @param keyword the keyword to filter the task list by.
+     * @return the filtered task list.
+     */
+    public ArrayList<Task> filterDescByKeyword(String keyword) {
+        ArrayList<Task> filteredTaskList = new ArrayList<>();
+        this.taskList.forEach((task -> {
+            if (task.description.contains(keyword)) {
+                filteredTaskList.add(task);
+            }
+        }));
+
+        return filteredTaskList;
+    }
+
+    /**
+     * Reverts the changes after an edit operation ({@code addTask}, {@code deleteTask}, {@code markTask},
+     * {@code unmarkTask}).
+     * <p>
+     * To use this method, the instance variable {@code previousTaskList} must be set
+     * to the state of {@code taskList} using {@code this.previousTaskList = this.getAllTasks();}.
+     * {@code previousTaskList} can be cleared using {@code commitChanges} below.
+     *
+     * @see #taskList
+     * @see #previousTaskList
+     * @see #commitChanges()
+     */
+    public void revertChanges() {
+        if (this.previousTaskList != null) {
+            this.taskList = new ArrayList<>(this.previousTaskList);
+            this.previousTaskList = null;
+        }
+    }
```

Comment:
Seems like it depends on client using it correctly (after an edit operation). Maybe can return boolean value depending on if there is a previous change to revert to?